File: 1267-count-servers-that-communicate/1267-count-servers-that-communicate.py
class Solution:
    def countServers(self, grid: List[List[int]]) -> int:
        rows = len(grid)
        cols = len(grid[0])
        
        # Check the input: size of the grid. If it is 0 then exit
        if not (cols and rows):
            return 0

        connected = 0                  # store number of connected computers                 
        points=[]                      # store coordinates of all the computers 
        comps_per_row = [0] * rows     # store number of computers in given row
        comps_per_col = [0] * cols     # store number of computers in given column
        
        for row_i in range(rows):
            for col_i in range(cols):
                if grid[row_i][col_i]:              # checking if given cell is not 0
                    points.append((row_i,col_i))    # add coordinated to the set
                    comps_per_row[row_i]+=1         # increase nimber of computers in a given row
                    comps_per_col[col_i]+=1         # increase nimber of computers in a given column 
        
        # iterate through all computers
        for row_i,col_i in points:
            # is there more than 1 computer in given row or column
            if comps_per_row[row_i]>1 or comps_per_col[col_i]>1 :
                # if yes, then computer is connected, count him      
                connected += 1                      
        
        return connected

File: 1267-count-servers-that-communicate/1267-count-servers-that-communicate.java
class Solution {
    public int countServers(int[][] grid) {
        //check for null input 
        if(grid == null || grid.length ==0)return 0;
        int rows = grid.length;
        int cols = grid[0].length;
        int[] rowCount = new int[rows];
        int[] colCount = new int[cols];
        int total = 0;
        //get the total number of servers on each row and col and in total
        for(int r=0; r< rows; r++) {
            for(int c=0; c< cols; c++) {
                if(grid[r][c] ==1){
                    total++;
                    rowCount[r]++;
                    colCount[c]++;
                }
            }
        }
        for(int r=0; r< rows; r++){
            for(int c=0; c< cols; c++) {
                if(grid[r][c] ==1){//if there is a server
                    //if this is the only server on this row and this col(so it has no neighbor)
                    if(rowCount[r] ==1 && colCount[c] ==1) total--;
                }
            }
        }   
        return total;
    }
}
//Time complexity: O(r*c)
//space complexity: O(Max(r,c))